session/schema_amender.go

```diff
@@ -0,0 +1,532 @@
+		colStr += fmt.Sprintf(" %s ", colInfo.Name)
+	}
+	colStr += "]"
+	res := fmt.Sprintf("AmenedOpType=%d phyTblID=%d idxID=%d columns=%v", a.AmendOpType, a.indexInfoAtCommit.Meta().ID,
+		a.indexInfoAtCommit.Meta().ID, colStr)
+	return res
+}
+
+func (a *amendOperationDeleteOldIndex) genMutations(ctx context.Context, sctx sessionctx.Context,
+	commitMutations tikv.CommitterMutations, kvMap *rowKvMap, resAddMutations *tikv.CommitterMutations) error {
+	for i, key := range commitMutations.GetKeys() {
+		keyOp := commitMutations.GetOps()[i]
+		if tablecodec.IsIndexKey(key) || tablecodec.DecodeTableID(key) != a.info.tblInfoAtCommit.Meta().ID {
+			continue
+		}
+		if !isDeleteOp(keyOp) {
+			continue
+		}
+		err := a.processRowKey(ctx, sctx, key, kvMap.oldRowKvMap, resAddMutations)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *amendOperationAddNewIndex) genMutations(ctx context.Context, sctx sessionctx.Context, commitMutations tikv.CommitterMutations,
+	kvMap *rowKvMap, resAddMutations *tikv.CommitterMutations) error {
+	for i, key := range commitMutations.GetKeys() {
+		keyOp := commitMutations.GetOps()[i]
+		if tablecodec.IsIndexKey(key) || tablecodec.DecodeTableID(key) != a.info.tblInfoAtCommit.Meta().ID {
+			continue
+		}
+		if !isInsertOp(keyOp) {
+			continue
+		}
+		err := a.processRowKey(ctx, sctx, key, kvMap.newRowKvMap, resAddMutations)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *amendOperationAddIndexInfo) genIndexKeyValue(ctx context.Context, sctx sessionctx.Context, kvMap map[string][]byte,
+	key []byte, kvHandle kv.Handle, keyOnly bool) ([]byte, []byte, error) {
+	chk := a.chk
+	chk.Reset()
+	val, ok := kvMap[string(key)]
+	if !ok {
+		// The Op_Put may not exist in old value kv map.
+		if keyOnly {
+			return nil, nil, nil
+		}
+		return nil, nil, errors.Errorf("key=%v is not found in new row kv map", kv.Key(key).String())
+	}
+	err := executor.DecodeRowValToChunk(sctx, a.schemaAndDecoder.schema, a.tblInfoAtStart.Meta(), kvHandle, val, chk, a.schemaAndDecoder.decoder)
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend decode value to chunk failed", zap.Error(err))
+		return nil, nil, errors.Trace(err)
+	}
+	idxVals := make([]types.Datum, 0, len(a.indexInfoAtCommit.Meta().Columns))
+	for _, oldCol := range a.relatedOldIdxCols {
+		idxVals = append(idxVals, chk.GetRow(0).GetDatum(oldCol.Offset, &oldCol.FieldType))
+	}
+
+	// Generate index key buf.
+	newIdxKey, distinct, err := tablecodec.GenIndexKey(sctx.GetSessionVars().StmtCtx,
+		a.tblInfoAtCommit.Meta(), a.indexInfoAtCommit.Meta(), a.tblInfoAtCommit.Meta().ID, idxVals, kvHandle, nil)
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend generate index key failed", zap.Error(err))
+		return nil, nil, errors.Trace(err)
+	}
+	if keyOnly {
+		return newIdxKey, []byte{}, nil
+	}
+
+	// Generate index value buf.
+	var containsNonBinaryString bool
+	for _, idxCol := range a.indexInfoAtCommit.Meta().Columns {
```

Comment:
Could we use `checkContainNonBinaryString()` in tables/index.go?

Reply:
Some refactoring is done to wrap the check method.